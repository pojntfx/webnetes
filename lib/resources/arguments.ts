import {
  API_VERSION,
  EResourceKind,
  IResource,
  IResourceMetadata,
} from "./resource";

export interface IArgumentsSpec {
  argv: string[];
}

export class Arguments implements IResource<IArgumentsSpec> {
  apiVersion = API_VERSION;
  kind = EResourceKind.ARGUMENTS;

  constructor(
    public metadata: IResourceMetadata,
    public spec: IArgumentsSpec
  ) {}
}
