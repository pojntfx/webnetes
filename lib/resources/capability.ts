import {
  API_VERSION,
  EResourceKind,
  IResource,
  IResourceMetadata,
} from "./resource";

export interface ICapabilitySpec {}

export class Capability implements IResource<ICapabilitySpec> {
  apiVersion = API_VERSION;
  kind = EResourceKind.CAPABILITY;

  constructor(
    public metadata: IResourceMetadata,
    public spec: ICapabilitySpec
  ) {}
}
