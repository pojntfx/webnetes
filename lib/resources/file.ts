import {
  API_VERSION,
  EResourceKind,
  IResource,
  IResourceMetadata,
} from "./resource";

export interface IFileSpec {
  repository: string;
  uri: string;
}

export class File implements IResource<IFileSpec> {
  apiVersion = API_VERSION;
  kind = EResourceKind.FILE;

  constructor(public metadata: IResourceMetadata, public spec: IFileSpec) {}
}
