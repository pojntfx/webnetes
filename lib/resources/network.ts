import {
  API_VERSION,
  EResourceKind,
  IResource,
  IResourceMetadata,
} from "./resource";

export interface INetworkSpec {
  signaler: string;
  stunServers: string[];
  turnServers: string[];
}

export class Network implements IResource<INetworkSpec> {
  apiVersion = API_VERSION;
  kind = EResourceKind.NETWORK;

  constructor(public metadata: IResourceMetadata, public spec: INetworkSpec) {}
}
