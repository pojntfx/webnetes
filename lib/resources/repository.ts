import {
  API_VERSION,
  EResourceKind,
  IResource,
  IResourceMetadata,
} from "./resource";

export interface IRepositorySpec {
  trackers: string[];
  stunServers: string[];
  turnServers: string[];
}

export class Repository implements IResource<IRepositorySpec> {
  apiVersion = API_VERSION;
  kind = EResourceKind.REPOSITORY;

  constructor(
    public metadata: IResourceMetadata,
    public spec: IRepositorySpec
  ) {}
}
