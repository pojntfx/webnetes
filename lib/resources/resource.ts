export const API_VERSION = "schema.webnetes.dev/v1alpha1";

export enum EResourceKind {
  RUNTIME = "Runtime",
  ARGUMENTS = "Arguments",
  CAPABILITY = "Capability",
  FILE = "File",
  NETWORK = "Network",
  PROCESSOR = "Processor",
  REPOSITORY = "Repository",
  SIGNALER = "Signaler",
  STUNSERVER = "StunServer",
  NETWORK_INTERFACE = "NetworkInterface",
  TRACKER = "Tracker",
  TURNSERVER = "TurnServer",
  WORKLOAD = "Workload",
  COORDINATES = "Coordinates",
  BENCHMARK_SCORE = "BenchmarkScore",
  PUBLIC_IP = "PublicIP",
}

export interface IResourceMetadata {
  name?: string;
  label: string;
}

export interface IResource<T> {
  apiVersion: string;
  kind: EResourceKind;
  metadata: IResourceMetadata;
  spec: T;
}
