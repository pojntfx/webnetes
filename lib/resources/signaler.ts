import {
  API_VERSION,
  EResourceKind,
  IResource,
  IResourceMetadata,
} from "./resource";

export interface ISignalerSpec {
  urls: string[];
  retryAfter: number;
}

export class Signaler implements IResource<ISignalerSpec> {
  apiVersion = API_VERSION;
  kind = EResourceKind.SIGNALER;

  constructor(public metadata: IResourceMetadata, public spec: ISignalerSpec) {}
}
