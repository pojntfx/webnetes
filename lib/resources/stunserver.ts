import {
  API_VERSION,
  EResourceKind,
  IResource,
  IResourceMetadata,
} from "./resource";

export interface IStunServerSpec {
  urls: string[];
}

export class StunServer implements IResource<IStunServerSpec> {
  apiVersion = API_VERSION;
  kind = EResourceKind.STUNSERVER;

  constructor(
    public metadata: IResourceMetadata,
    public spec: IStunServerSpec
  ) {}
}
