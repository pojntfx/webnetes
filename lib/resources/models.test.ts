import { Arguments } from "./arguments";
import { Capability } from "./capability";
import { File } from "./file";
import { Network } from "./network";
import { Processor } from "./processor";
import { Repository } from "./repository";
import { EResourceKind, IResource } from "./resource";
import { Runtime } from "./runtime";
import { Signaler } from "./signaler";
import { StunServer } from "./stunserver";
import { Subnet } from "./subnet";
import { Tracker } from "./tracker";
import { TurnServer } from "./turnserver";
import { Workload } from "./workload";

const argumentsData = require("./arguments.data.json");
const capabilityData = require("./capability.data.json");
const fileData = require("./file.data.json");
const networkData = require("./network.data.json");
const processorData = require("./processor.data.json");
const repositoryData = require("./repository.data.json");
const runtimeData = require("./runtime.data.json");
const signalerData = require("./signaler.data.json");
const stunserverData = require("./stunserver.data.json");
const subnetData = require("./subnet.data.json");
const trackerData = require("./tracker.data.json");
const turnserverData = require("./turnserver.data.json");
const workloadData = require("./workload.data.json");

describe("Models", () => {
  test.each([
    [argumentsData],
    [capabilityData],
    [fileData],
    [networkData],
    [processorData],
    [repositoryData],
    [runtimeData],
    [signalerData],
    [stunserverData],
    [subnetData],
    [trackerData],
    [turnserverData],
    [workloadData],
  ])("%s should be %s", (resource: IResource<any>) => {
    let actual: any = {};

    switch (resource.kind) {
      case EResourceKind.ARGUMENTS: {
        actual = new Arguments(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.CAPABILITY: {
        actual = new Capability(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.FILE: {
        actual = new File(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.NETWORK: {
        actual = new Network(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.PROCESSOR: {
        actual = new Processor(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.REPOSITORY: {
        actual = new Repository(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.RUNTIME: {
        actual = new Runtime(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.SIGNALER: {
        actual = new Signaler(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.STUNSERVER: {
        actual = new StunServer(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.SUBNET: {
        actual = new Subnet(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.TRACKER: {
        actual = new Tracker(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.TURNSERVER: {
        actual = new TurnServer(resource.metadata, resource.spec);

        break;
      }

      case EResourceKind.WORKLOAD: {
        actual = new Workload(resource.metadata, resource.spec);

        break;
      }
    }

    expect(actual).toMatchSnapshot();
  });
});
