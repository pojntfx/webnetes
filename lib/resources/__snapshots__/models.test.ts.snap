// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Arguments',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Arguments {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Arguments",
  "metadata": Object {
    "label": "echo_server",
    "name": "Echo Server Configuration",
  },
  "spec": Object {
    "argv": Array [
      "-laddr",
      "127.0.0.1:1234",
    ],
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Capability',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Capability {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Capability",
  "metadata": Object {
    "label": "bind_alias",
    "name": "Binding aliases",
  },
  "spec": Object {
    "privileged": true,
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'File',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
File {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "File",
  "metadata": Object {
    "label": "echo_server",
    "name": "Echo Server Binary",
  },
  "spec": Object {
    "repository": "webtorrent_public",
    "uri": "magnet:?xt=urn:btih:3e6e88fd7d52ad0a47bbb7c10e141573f9b5608e",
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Network',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Network {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Network",
  "metadata": Object {
    "label": "unisockets_public",
    "name": "Public unisockets network",
  },
  "spec": Object {
    "signaler": "unisockets_public",
    "stunServers": Array [
      "google",
      "twillio",
    ],
    "turnServers": Array [
      "twillio",
    ],
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Processor',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Processor {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Processor",
  "metadata": Object {
    "label": "felicitass_iphone",
    "name": "Felicitas's iPhone",
  },
  "spec": Object {
    "capabilities": Array [
      "bind_alias",
      "connect_to_alias",
    ],
    "runtimes": Array [
      "generic_wasi",
    ],
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Repository',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Repository {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Repository",
  "metadata": Object {
    "label": "webtorrent_public",
    "name": "Public WebTorrent",
  },
  "spec": Object {
    "trackers": Array [
      "openwebtorrent",
      "btorrent",
      "fastcast",
    ],
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Runtime',
  metadata: [Object]
} should be %s 1`] = `
Runtime {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Runtime",
  "metadata": Object {
    "label": "generic_wasi",
    "name": "Generic WASI",
  },
  "spec": undefined,
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Signaler',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Signaler {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Signaler",
  "metadata": Object {
    "label": "unisockets_public",
    "name": "Public unisockets Signaling Server",
  },
  "spec": Object {
    "retryAfter": 1000,
    "urls": Array [
      "wss://unisockets.herokuapp.com",
    ],
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'StunServer',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
StunServer {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "StunServer",
  "metadata": Object {
    "label": "google",
    "name": "Google STUN Server",
  },
  "spec": Object {
    "urls": Array [
      "stun.l.google.com:19302",
    ],
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Subnet',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Subnet {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Subnet",
  "metadata": Object {
    "label": "echo_network",
    "name": "Echo Network",
  },
  "spec": Object {
    "network": "unisockets_public",
    "prefix": "127.0.0",
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Tracker',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Tracker {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Tracker",
  "metadata": Object {
    "label": "openwebtorrent",
    "name": "OpenWebTorrent",
  },
  "spec": Object {
    "urls": Array [
      "wss://tracker.openwebtorrent.com",
    ],
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'TurnServer',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
TurnServer {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "TurnServer",
  "metadata": Object {
    "label": "twillio",
    "name": "Twillio TURN Server",
  },
  "spec": Object {
    "credential": "mypass=",
    "urls": Array [
      "turn:global.turn.twilio.com:3478?transport=tcp",
    ],
    "username": "f4b4035eaa76f4a55de5f4351567",
  },
}
`;

exports[`Models {
  apiVersion: 'webnetes.felicitas.pojtinger.com/v1alpha1',
  kind: 'Workload',
  metadata: [Object],
  spec: [Object]
} should be %s 1`] = `
Workload {
  "apiVersion": "webnetes.felicitas.pojtinger.com/v1alpha1",
  "kind": "Workload",
  "metadata": Object {
    "label": "echo_server",
    "name": "Echo Server",
  },
  "spec": Object {
    "arguments": "echo_server",
    "capabilities": Array [
      "bind_alias",
    ],
    "file": "echo_server",
    "runtime": "generic_wasi",
    "subnet": "echo_network",
  },
}
`;
